File: plugins/lofi/vinyl/vinyl.js
// Vinyl — old-record texture: pink-noise hiss plus randomly-triggered
// crackle pops, mixed into the signal. Pink noise comes from a small
// Voss-McCartney generator (5 octave-band white sources, each refreshed
// at a different rate keyed off the running counter's lowest set bit).
// Crackles fire at a Poisson rate scaled by the knob and decay through
// a one-pole envelope so they have an analog "snap" rather than a click.

function VinylProcessor() {
  this.pink = [0, 0, 0, 0, 0];
  this.counter = 0;
  this.crackleState = 0;
}
VinylProcessor.parameterDescriptors = [
  { name: 'hiss',    label: 'Hiss',    defaultValue: 0.12, minValue: 0, maxValue: 1 },
  { name: 'crackle', label: 'Crackle', defaultValue: 0.30, minValue: 0, maxValue: 1 },
  { name: 'mix',     label: 'Mix',     defaultValue: 1,    minValue: 0, maxValue: 1 }
];
VinylProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var hiss = parameters.hiss[0];
  var crackle = parameters.crackle[0];
  var mix = parameters.mix[0];
  for (var i = 0; i < input.length; i++) {
    this.counter = (this.counter + 1) | 0;
    // Voss-McCartney: refresh band whose bit position equals the
    // lowest set bit of the counter — gives 1/f noise spectrum.
    var lsb = 0;
    var c = this.counter;
    while ((c & 1) === 0 && lsb < this.pink.length - 1) {
      lsb++;
      c = c >> 1;
    }
    this.pink[lsb] = Math.random() * 2 - 1;
    var pinkSum = 0;
    for (var p = 0; p < this.pink.length; p++) pinkSum += this.pink[p];
    pinkSum *= 0.18;

    if (Math.random() < crackle * 0.0008) {
      this.crackleState = (Math.random() * 2 - 1) * 0.8;
    }
    this.crackleState *= 0.85;

    var dirt = pinkSum * hiss * 0.4 + this.crackleState * crackle;
    output[i] = input[i] + dirt * mix;
  }
  return true;
};

registerProcessor('vinyl', VinylProcessor);

File: plugins/lofi/vinyl/Vinyl.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;
import java.util.Random;

// Vinyl (native port) — pink noise hiss (Voss-McCartney) plus
// Poisson-triggered crackle pops with one-pole decay. Aged-record
// texture under any take.
public final class Vinyl implements VocalMonitorNativePlugin {
    private final float[] pink = new float[5];
    private int counter = 0;
    private float crackleState = 0f;
    private final Random rng = new Random();
    private float hiss = 0.12f, crackle = 0.30f, mix = 1f;

    @Override
    public void init(int sr) {
        for (int i = 0; i < 5; i++) pink[i] = 0f;
        counter = 0;
        crackleState = 0f;
    }

    @Override public String[] parameterNames() { return new String[] { "hiss", "crackle", "mix" }; }
    @Override public float parameterMin(String n) { return 0f; }
    @Override public float parameterMax(String n) { return 1f; }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "hiss":    return 0.12f;
            case "crackle": return 0.30f;
            case "mix":     return 1f;
            default:        return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "hiss":    return "Hiss";
            case "crackle": return "Crackle";
            case "mix":     return "Mix";
            default:        return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "hiss":    hiss = v; break;
            case "crackle": crackle = v; break;
            case "mix":     mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float hissScaled = hiss * 0.4f;
        final float crackleLocal = crackle;
        final float crackleProb = crackleLocal * 0.0008f;
        final float mixLocal = mix;
        final Random r = rng;
        final float[] pk = pink;
        int c = counter;
        float cs = crackleState;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            c++;
            int lsb = 0;
            int cc = c;
            while ((cc & 1) == 0 && lsb < pk.length - 1) {
                lsb++;
                cc >>= 1;
            }
            pk[lsb] = r.nextFloat() * 2f - 1f;
            float pinkSum = (pk[0] + pk[1] + pk[2] + pk[3] + pk[4]) * 0.18f;

            if (r.nextFloat() < crackleProb) {
                cs = (r.nextFloat() * 2f - 1f) * 0.8f;
            }
            cs *= 0.85f;

            float dirt = pinkSum * hissScaled + cs * crackleLocal;
            output[i] = input[i] + dirt * mixLocal;
        }
        counter = c;
        crackleState = cs;
    }
}
